ConsulAndServiceDiscovery/README.md

```diff
@@ -0,0 +1,114 @@
+# Introduction
+
+This document describes the consul usage and service discovery guidelines. This
+document is sure to evolve as we gain more experience with running consul in
+production. If you have any questions please use the Wikia #consul chat channel.
+
+## Introduction to Consul and Service Discovery
+
+For an introduction to consul please read the [documentation
+provided](https://www.consul.io/intro/) by the creators of consul. Read that
+first.
+
+Services can be "discovered" via DNS or the HTTP API. As an example, consider a
+query for a discussion service node. When using the DNS interface one would
+query for `discussion.service.sjc.consul`. Below is an example:
+
+```
+dig @localhost -p 8600 discussion.service.sjc.consul
+```
+
+The answer section of the above query will provide the IP addresses of instances
+that have registered themselves as `discussion` "services" and are healthy. 
+
+If you want to also query for the port that the service is listening one
+add `SRV` to the end of the above `dig` query.
+
+Consul DNS addresses take the form `TAG.SERVICE.service.DATACENTER.consul`.
+
+A similar request for "healthy" instances of the discussion service could query
+the HTTP API using the following:
+
+```
+curl 'http://consul.service.sjc.consul:8500/v1/health/service/discussion?passing'
+```
+
+The `passing` query parameter will limit the output to "healthy" services.
+
+# Guidelines
+
+## DNS Schema
+
+### `TAG`
+
+Tags are reserved for attributes that are specific to a set of registered
+instances of a service. Below are some examples of how tags might be used to
+distinguish between instances.
+
+ * `api` this instance of the service provides an HTTP API to access the
+   service.
+ * `admin.api` [note that the tag includes the `.`] identifies an administrative
+   API that is not exposed to the public.
+ * `db` this instance provides persistence for the service.
```

Comment:
> I think [tag].db-servicename[.dc].consul is a good standard (although
servicename-db would sort better), since the database is ultimately a
different service, not just a variant of the same service.

This is what we ended up settling on. See
https://github.com/Wikia/guidelines/tree/master/ConsulAndServiceDiscovery#service
.

On Wed, Jul 29, 2015 at 9:06 AM, Charles Miller <notifications@github.com>
wrote:

> In ConsulAndServiceDiscovery/README.md
> <https://github.com/Wikia/guidelines/pull/86#discussion_r35777467>:
>
> > +
> > +# Guidelines
> > +
> > +## DNS Schema
> > +
> > +### `TAG`
> > +
> > +Tags are reserved for attributes that are specific to a set of registered
> > +instances of a service. Below are some examples of how tags might be used to
> > +distinguish between instances.
> > +
> > + * `api` this instance of the service provides an HTTP API to access the
> > +   service.
> > + * `admin.api` [note that the tag includes the `.`] identifies an administrative
> > +   API that is not exposed to the public.
> > + * `db` this instance provides persistence for the service.
>
> I think [tag].db-servicename[.dc].consul is a good standard (although
> servicename-db would sort better), since the database is ultimately a
> different service, not just a variant of the same service.
>
> If we think having multiple names for the same IP/port combo might cause
> problems in Consul, I'd rather we found out for sure instead of ditching a
> potentially very useful solution because it "might" cause problems.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Wikia/guidelines/pull/86/files#r35777467>.
>